core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt

```diff
@@ -18,16 +21,69 @@ object UnitActionModifiers {
             .filter { unique -> unique.conditionals.none { it.type == UniqueType.UnitActionExtraLimitedTimes } }
             .filter { canUse(unit, it) }
 
-    private fun getMovementPointsToUse(actionUnique: Unique): Int {
+    private fun getMovementPointsToUse(unit: MapUnit, actionUnique: Unique, defaultAllMovement: Boolean = false): Int {
+        if (actionUnique.conditionals.any { it.type == UniqueType.UnitActionMovementCostAll })
+            return unit.getMaxMovement()
         val movementCost = actionUnique.conditionals
             .filter { it.type == UniqueType.UnitActionMovementCost }
             .minOfOrNull { it.params[0].toInt() }
-        if (movementCost != null) return movementCost
-        return 1
+        val movementCostRequired = actionUnique.conditionals
+            .filter { it.type == UniqueType.UnitActionMovementCostRequired }
+            .minOfOrNull { it.params[0].toInt() }
+        if (movementCost != null && movementCostRequired != null) {
+            return maxOf(movementCost, movementCostRequired)
+        } else {
+            if (movementCost != null)
+                return movementCost
+            if (movementCostRequired != null)
+                return movementCostRequired
+        }
+        return if (defaultAllMovement) unit.getMaxMovement() else 1
+    }
+
+    private fun getMovementPointsRequired(actionUnique: Unique): Int {
+        if (actionUnique.conditionals.any { it.type == UniqueType.UnitActionMovementCostAll })
+            return 1
+        val movementCostRequired = actionUnique.conditionals
+            .filter { it.type == UniqueType.UnitActionMovementCostRequired }
+            .minOfOrNull { it.params[0].toInt() }
+        return movementCostRequired ?: 1
+    }
+
+    /**Check if the stat costs in this Action Modifier can be spent by the Civ/Closest City without
+     * going into the negatives
+     * @return Boolean
+     */
+    private fun canSpendStatsCost(unit: MapUnit, actionUnique: Unique): Boolean {
+        for (conditional in actionUnique.conditionals.filter { it.type == UniqueType.UnitActionStatsCost }) {
+            for ((stat, value) in conditional.stats) {
+                if (stat in Stat.statsWithCivWideField) {
+                    if (!unit.civ.hasStatToBuy(stat, value.toInt()))
+                        return false
+                } else {
+                    if (unit.getClosestCity() != null) {
```

Comment:
Val closestCity = ..., Reverse if to reduce nesting - this tree is way too complex!

Reply:
Yeah just realized since I added functionality to the City I can simplify this

Reply:
Done